src/v/iceberg/conversion/values_json.cc

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2025 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "iceberg/conversion/values_json.h"
+
+#include "bytes/iobuf_parser.h"
+#include "iceberg/conversion/ir_json.h"
+#include "iceberg/conversion/time_rfc3339.h"
+#include "iceberg/values.h"
+#include "serde/json/parser.h"
+
+#include <memory>
+#include <optional>
+#include <variant>
+
+namespace iceberg {
+
+namespace {
+
+std::optional<primitive_value> convert_primitive(
+  experimental::serde::json::parser& p, const primitive_type& ft) {
+    using token = experimental::serde::json::token;
+
+    switch (p.token()) {
+    case token::value_null:
+        return std::nullopt;
+    case token::value_true:
+        if (!std::holds_alternative<boolean_type>(ft)) {
+            throw value_conversion_exception(
+              "Miss-match primitive type boolean value");
+        }
+        return iceberg::boolean_value(true);
+    case token::value_false:
+        if (!std::holds_alternative<boolean_type>(ft)) {
+            throw value_conversion_exception(
+              "Miss-match primitive type boolean value");
+        }
+        return iceberg::boolean_value(false);
+    case token::value_int:
+        if (!std::holds_alternative<long_type>(ft)) {
+            throw value_conversion_exception(
+              "Miss-match primitive type integer value");
+        }
+        return iceberg::long_value(p.value_int());
+    case token::value_double:
+        if (!std::holds_alternative<double_type>(ft)) {
+            throw value_conversion_exception(
+              "Miss-match primitive type double value");
+        }
+        return iceberg::double_value(p.value_double());
+    case token::value_string: {
+        // Trivial case for string type.
+        if (std::holds_alternative<string_type>(ft)) {
+            return iceberg::string_value(p.value_string());
+        }
+
+        auto linearize = [](iobuf buf, size_t max_bytes) {
+            if (buf.size_bytes() > max_bytes) {
+                throw value_conversion_exception(fmt::format(
+                  "String value exceeds maximum length of {} bytes: {}",
+                  max_bytes,
+                  buf.size_bytes()));
+            }
+
+            iobuf_parser iop(std::move(buf));
```

Comment:
there might be a slight, tiny, possibly non-existent performance advantage here if you used `const_iobuf_parser`